c/rce.h

```diff
@@ -0,0 +1,382 @@
+#ifndef XUDT_RCE_SIMULATOR_C_RCE_H_
+#define XUDT_RCE_SIMULATOR_C_RCE_H_
+#include "ckb_smt.h"
+#include "xudt_rce_mol.h"
+#include "xudt_rce_mol2.h"
+
+enum ErrorCode {
+  // 0 is the only success code. We can use 0 directly.
+
+  // inherit from simple_udt
+  ERROR_ARGUMENTS_LEN = -1,
+  ERROR_ENCODING = -2,
+  ERROR_SYSCALL = -3,
+  ERROR_SCRIPT_TOO_LONG = -21,
+  ERROR_OVERFLOWING = -51,
+  ERROR_AMOUNT = -52,
+
+  // error code is starting from 40, to avoid conflict with
+  // common error code in other scripts.
+  ERROR_CANT_LOAD_LIB = 40,
+  ERROR_CANT_FIND_SYMBOL,
+  ERROR_INVALID_RCE_ARGS,
+  ERROR_NOT_ENOUGH_BUFF,
+  ERROR_INVALID_FLAG,
+  ERROR_INVALID_ARGS_FORMAT,
+  ERROR_INVALID_WITNESS_FORMAT,
+  ERROR_INVALID_MOL_FORMAT,
+  ERROR_BLAKE2B_ERROR,
+  ERROR_HASH_MISMATCHED,
+  ERROR_RCRULES_TOO_DEEP,
+  ERROR_TOO_MANY_RCRULES,
+  ERROR_RCRULES_PROOFS_MISMATCHED,
+  ERROR_SMT_VERIFY_FAILED,
+  ERROR_RCE_EMERGENCY_HALT,
+  ERROR_NOT_VALIDATED,
+  ERROR_TOO_MANY_LOCK,
+  ERROR_ON_BLACK_LIST,
+  ERROR_TYPE_FREEZED,
+  ERROR_APPEND_ONLY,
+  ERROR_EOF,
+  ERROR_TOO_LONG_PROOF,
+};
+
+#define CHECK2(cond, code) \
+  do {                     \
+    if (!(cond)) {         \
+      err = code;          \
+      ASSERT(0);           \
+      goto exit;           \
+    }                      \
+  } while (0)
+
+#define CHECK(_code)    \
+  do {                  \
+    int code = (_code); \
+    if (code != 0) {    \
+      err = code;       \
+      ASSERT(0);        \
+      goto exit;        \
+    }                   \
+  } while (0)
+
+int get_extension_data(uint32_t index, uint8_t* buff, uint32_t buff_len,
+                       uint32_t* out_len);
+
+#define BLAKE2B_BLOCK_SIZE 32
+#define MAX_EXTENSION_DATA_SIZE 32768
+#define MAX_LOCK_SCRIPT_HASH_COUNT 2048
+#define MAX_RCRULES_COUNT 8192
+#define MAX_RECURSIVE_DEPTH 16
+#define MAX_TEMP_PROOF_LENGTH 32768
+
+// RC stands for Regulation Compliance
+typedef struct RCRule {
+  uint8_t smt_root[32];
+  uint8_t flags;
+} RCRule;
+
+typedef struct RceState {
+  RCRule rcrules[MAX_RCRULES_COUNT];
+  uint32_t rcrules_count;
+} RceState;
+
+void rce_init_state(RceState* state) { state->rcrules_count = 0; }
+
+// molecule doesn't provide names
+typedef enum RCDataUnionType {
+  RCDataUnionRule = 0,
+  RCDataUnionCellVec = 1
+} RCDataUnionType;
+
+// RCE scripts leverage optimized sparse merkle tree
+// (https://github.com/jjyr/sparse-merkle-tree)(SMT) extensively to reduce
+// storage costs. For each sparse merkle tree used here, the key will be lock
+// script hash, values are either 0 or 1: 0 represents the corresponding lock
+// hash is missing in the sparse merkle tree, whereas 1 means the lock hash is
+// included in the sparse merkle tree.
+uint8_t SMT_VALUE_NOT_EXISTING[SMT_VALUE_BYTES] = {0};
+uint8_t SMT_VALUE_EXISTING[SMT_VALUE_BYTES] = {1};
```

Comment:
The value should be ` 0x0000 ... 0001` according to [RCE RFC](https://talk.nervos.org/t/rfc-regulation-compliance-extension/5338), but here the value is `0x0100 ... 0000`.

Reply:
uint8_t SMT_VALUE_EXISTING[SMT_VALUE_BYTES] = {[31]=1};
https://en.cppreference.com/w/c/language/array_initialization